### PR TITLE
CA2241: AutoDetectStringFormattingMethods

### PR DESCRIPTION
**Pull request by manfred-brands**
In the end this was very simple as the method to detect a Format method already existed.

We could completely delete the existing configuration and simply check if the called method has a format parameter followed by a params parameter. This can not be much more expensive than checking the configuration.

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs [hunk 1]
```diff
@@ -355,6 +355,13 @@ public StringFormatInfo(Compilation compilation)
                     return info;
                 }
 
+                // Check if method could be a string formatting method.
+                // This is defined as a method with a 'format' argument followed by an 'params object[]'.
+                if (TryGetFormatInfo(method, out info))
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs:360]: As far as I recall, this behavior was considered in the original implementation, but was not chosen due to following reasons:
1. Potential of being a breaking change for the rule.
2. Makes this a heuristic based rule. Current implementation of the rule is conservative to the known or user configured set of formatting methods. This makes it aggressive to fire on all methods which have similar signature.

Given that this rule is marked as `RuleLevel.BuildWarningCandidate`, I am reluctant to change the default behavior to make it so aggressive. At the minimum, we need an option to guard this new behavior so users can select the preferred level of aggressiveness.

@dotnet/roslyn-analysis for opinions.

**Comment by manfred-brands** [comment]
I added a commit which makes this behaviour configurable.
In addition the format parameter is now verified to be a string.

**Comment by mavasani** [comment]
Change looks good to me overall with the default behavior being unchanged. Please add relevant documentation. Also we probably want a better option name, but I cannot think of a good one. Any suggestions?

**Comment by manfred-brands** [comment]
Renamed option to: try_determine_additional_string_formatting_methods_automatically. Should be clear.
Added documentation and addressed your other review comments.

File: docs/Analyzer Configuration.md [hunk 2]
```diff
@@ -239,6 +239,16 @@ Examples:
 |`dotnet_code_quality.additional_string_formatting_methods = NS.MyType.MyFormat(ParamType)` | Matches specific method 'MyFormat' with given fully qualified signature
 |`dotnet_code_quality.additional_string_formatting_methods = NS1.MyType1.MyFormat1(ParamType)\|NS2.MyType2.MyFormat2(ParamType)` | Matches specific methods 'MyFormat1' and 'MyFormat2' with respective fully qualified signature
 
+Option Name: `try_determine_additional_string_formatting_methods_automatically`
```
Comment (mavasani) [inline docs/Analyzer Configuration.md:242]: Can you please add a more detailed description here on what this option does?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs [hunk 3]
```diff
@@ -355,6 +355,17 @@ public StringFormatInfo(Compilation compilation)
                     return info;
                 }
 
+                // Check if the user configured automatic determination of formatting methods.
+                // If so, check if the method called has a 'string format' parameter followed by an params array.
+                bool defaultTryDetermineAdditionalStringFormattingMethodsAutomatically = false;
+                var canTryDetermineAdditionalStringFormattingMethodsAutomatically = context.Options.GetBoolOptionValue(EditorConfigOptionNames.TryDetermineAdditionalStringFormattingMethodsAutomatically,
+                        Rule, context.Operation.Syntax.SyntaxTree, context.Compilation, defaultTryDetermineAdditionalStringFormattingMethodsAutomatically, context.CancellationToken);
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs:362]:
```suggestion
                var determineAdditionalStringFormattingMethodsAutomatically = context.Options.GetBoolOptionValue(EditorConfigOptionNames.TryDetermineAdditionalStringFormattingMethodsAutomatically,
                        Rule, context.Operation.Syntax.SyntaxTree, context.Compilation, defaultValue: false, context.CancellationToken);
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs [hunk 4]
```diff
@@ -355,6 +355,17 @@ public StringFormatInfo(Compilation compilation)
...
+                if (canTryDetermineAdditionalStringFormattingMethodsAutomatically &&
+                    TryGetFormatInfo(method, out info) && info.ExpectedStringFormatArgumentCount == -1)
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs:364]:
```suggestion
                if (determineAdditionalStringFormattingMethodsAutomatically  &&
                    TryGetFormatInfo(method, out info) &&
                    info.ExpectedStringFormatArgumentCount == -1)
```

File: src/Utilities/Compiler/Options/EditorConfigOptionNames.cs [hunk 5]
```diff
@@ -67,6 +67,13 @@ internal static partial class EditorConfigOptionNames
         /// </summary>
         public const string AdditionalStringFormattingMethods = "additional_string_formatting_methods";
 
+        /// <summary>
+        /// String option to enable heuristically detecting of additional string formatting methods for CA2241 (https://docs.microsoft.com/visualstudio/code-quality/ca2241).
```
Comment (mavasani) [inline src/Utilities/Compiler/Options/EditorConfigOptionNames.cs:71]: This exact comment can be added to the `Analyzer Configuration.md` doc section.

Comment (manfred-brands) [reply src/Utilities/Compiler/Options/EditorConfigOptionNames.cs:71]: I noticed that the comment is wrong, it should read 'Boolean option'.

File: docs/Analyzer Configuration.md (hunk 2 above)
Comment (manfred-brands) [reply docs/Analyzer Configuration.md:242]: Done.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs (hunk 3 above)
Comment (manfred-brands) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs]: Done.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs (hunk 4 above)
Comment (manfred-brands) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideCorrectArgumentsToFormattingMethods.cs]: Done.

**Comment by manfred-brands** [comment]
Changes made and pushed. Would you like for me to collapse these commits into one or will that confuse the automatic system?

**Comment by manfred-brands** [comment]
Thanks. Raised pull request for documentation: https://github.com/MicrosoftDocs/visualstudio-docs/pull/5391

**Pull request merged**